runtime/catalog-karapace/src/main/java/io/aklivity/zilla/runtime/catalog/karapace/internal/KarapaceCatalogHandler.java

```diff
@@ -97,21 +174,98 @@ public int resolve(
+                    }
+                }
+                catch (Throwable ex)
+                {
+                    existing.completeExceptionally(ex);
+                }
+            }
+            CompletableFuture<CachedSchemaId> future = cachedSchemaIds.merge(schemaKey, newFuture, (v1, v2) ->
+                v1.getNow(IN_PROGRESS).retry() &&
+                    (v1.getNow(IN_PROGRESS).id == NO_SCHEMA_ID || v1.getNow(IN_PROGRESS).expired(maxAgeMillis)) ? v2 : v1);
+            if (future == newFuture)
             {
-                schemaIds.put(checkSum, new CachedSchemaId(System.currentTimeMillis(), schemaId));
+                try
+                {
+                    String response = sendHttpRequest(MessageFormat.format(SUBJECT_VERSION_PATH, subject, version));
+                    if (response == null)
+                    {
+                        if (retryAttempts.getAndIncrement() == 0)
+                        {
+                            retryAfter = RETRY_INITIAL_DELAY_MS_DEFAULT;
+                            event.onUnretrievableSchemaSubjectVersion(catalogId, subject, version);
+                        }
+
+                        if (cachedSchemaId != null)
+                        {
+                            if (cachedSchemaId.retryAfter != RESET_RETRY_DELAY_MS_DEFAULT)
+                            {
+                                retryAfter = Math.min(cachedSchemaId.retryAfter << 1, maxAgeMillis);
+                            }
+                            newFuture.complete(new CachedSchemaId(cachedSchemaId.timestamp, cachedSchemaId.id,
+                                retryAttempts, retryAfter));
+                        }
+                        else
+                        {
+                            newFuture.complete(new CachedSchemaId(System.currentTimeMillis(), NO_SCHEMA_ID,
+                                retryAttempts, retryAfter));
+                        }
+                    }
+                    else if (response != null)
+                    {
+                        if (retryAttempts.getAndSet(0) > 0)
+                        {
+                            event.onRetrievableSchemaSubjectVersion(catalogId, subject, version);
+                        }
+                        newFuture.complete(new CachedSchemaId(System.currentTimeMillis(), request.resolveResponse(response),
+                            retryAttempts, retryAfter));
+                    }
+                }
+                catch (Throwable ex)
+                {
+                    newFuture.completeExceptionally(ex);
+                }
+            }
+            assert future != null;
+            try
+            {
+                cachedSchemaId = future.get();
+                schemaId = cachedSchemaId.id;
+                if (schemaId != NO_SCHEMA_ID)
+                {
+                    schemaIds.put(schemaKey, cachedSchemaId);
+                    if (cachedSchemaId.retryAttempts.compareAndSet(1, 2))
```

Comment:
This doesn't look right to me, because the value of `retryAttempts` should be independent of serving stale schema, right?